### ISSUE TITLE
New lint: `unnecessary_option`

### ISSUE THREAD

**Issue by jubnzv** [rust-lang/rust-clippy#11612]
### What it does

This lint would warn against using the `Option` type for local variables when the `None` value is unused.

### Advantage

Remove the unnecessary `Option`s  makes the source code more readable and effective.

### Drawbacks

* To make the lint more effective and simple, the implementation of the lint should not be exhaustive. It makes sense to iterate over HIR and mark an `Option` variable as "used" if it is used as an argument of a function call, i.e. don't propagate these values across functions.
* The lint will warn only at the definition point of the local variable. Therefore, the user has to fix all the cases when the `Some` value is used by himself.

### Example

```rust
fn test() {
  let a = Some(5);
  // Any statements here, but the `None` value of `a` is unused, e.g.:
  // if a.is_some() { /* ... */ }
  // if let Some(v) = a { /* ... */ }
}
```

Could be written as:

```rust
fn test() {
  let v = 5;
}
```
